vpr/src/place/net_cost_handler.cpp

```diff
@@ -0,0 +1,2277 @@
+        ClusterNetId net_id = ts_nets_to_update[inet_affected];
+
+        if (cube_bb) {
+            place_move_ctx.bb_coords[net_id] = ts_bb_coord_new[net_id];
+        } else {
+            place_move_ctx.layer_bb_coords[net_id] = layer_ts_bb_coord_new[net_id];
+        }
+
+        for (int layer_num = 0; layer_num < g_vpr_ctx.device().grid.get_num_layers(); layer_num++) {
+            place_move_ctx.num_sink_pin_layer[size_t(net_id)][layer_num] = ts_layer_sink_pin_count[size_t(net_id)][layer_num];
+        }
+
+        if (cluster_ctx.clb_nlist.net_sinks(net_id).size() >= SMALL_NET) {
+            if (cube_bb) {
+                place_move_ctx.bb_num_on_edges[net_id] = ts_bb_edge_new[net_id];
+            } else {
+                place_move_ctx.layer_bb_num_on_edges[net_id] = layer_ts_bb_edge_new[net_id];
+            }
+        }
+
+        net_cost[net_id] = proposed_net_cost[net_id];
+
+        /* negative proposed_net_cost value is acting as a flag. */
+        proposed_net_cost[net_id] = -1;
+        bb_updated_before[net_id] = NetUpdateState::NOT_UPDATED_YET;
+    }
+}
+
+void reset_move_nets(int num_nets_affected) {
+    /* Reset the net cost function flags first. */
+    for (int inet_affected = 0; inet_affected < num_nets_affected;
+         inet_affected++) {
+        ClusterNetId net_id = ts_nets_to_update[inet_affected];
+        proposed_net_cost[net_id] = -1;
+        bb_updated_before[net_id] = NetUpdateState::NOT_UPDATED_YET;
+    }
+}
+
+void recompute_costs_from_scratch(const t_placer_opts& placer_opts,
+                                const t_noc_opts& noc_opts,
+                                const PlaceDelayModel* delay_model,
+                                const PlacerCriticalities* criticalities,
+                                t_placer_costs* costs) {
+    auto check_and_print_cost = [](double new_cost,
+                                   double old_cost,
+                                   const std::string& cost_name) {
+        if (!vtr::isclose(new_cost, old_cost, ERROR_TOL, 0.)) {
+            std::string msg = vtr::string_fmt(
+                "in recompute_costs_from_scratch: new_%s = %g, old %s = %g, ERROR_TOL = %g\n",
+                cost_name.c_str(), new_cost, cost_name.c_str(), old_cost, ERROR_TOL);
+            VPR_ERROR(VPR_ERROR_PLACE, msg.c_str());
+        }
+    };
+
+    double new_bb_cost = recompute_bb_cost();
+    check_and_print_cost(new_bb_cost, costs->bb_cost, "bb_cost");
+    costs->bb_cost = new_bb_cost;
+
+    if (placer_opts.place_algorithm.is_timing_driven()) {
+        double new_timing_cost = 0.;
+        comp_td_costs(delay_model, *criticalities, &new_timing_cost);
+        check_and_print_cost(new_timing_cost, costs->timing_cost, "timing_cost");
+        costs->timing_cost = new_timing_cost;
+    } else {
+        VTR_ASSERT(placer_opts.place_algorithm == BOUNDING_BOX_PLACE);
+        costs->cost = new_bb_cost * costs->bb_cost_norm;
+    }
+
+    if (noc_opts.noc) {
```

Comment:
All these noc cost checks can be shortened by using vtr::isclose with an absolute tolerance of MIN_EXPECTED_NOC_LATENCY_COST etc. (i.e. set the last argument to that value).  So the if statements can be taken out, the check_and_print_cost lambda should take one more parameter (absolute cost threshold) and that parameter should be passed to vtr::isclose().

Reply:
Can file an issue to @soheilshahrouz if you want him to clean this up.